providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/grpc/GrpcConnector.java

```diff
@@ -1,174 +1,232 @@
-            final StreamObserver<EventStreamResponse> responseObserver = new EventStreamObserver(sync, this.cache,
-                    this::onConnectionEvent, backoff::shouldRetrySilently);
+    public void shutdown() throws InterruptedException {
+        log.info("Shutting down GRPC connection...");
+        if (reconnectExecutor != null) {
+            reconnectExecutor.shutdownNow();
+            reconnectExecutor.awaitTermination(deadline, TimeUnit.MILLISECONDS);
+        }
 
-            ServiceGrpc.ServiceStub localServiceStub = this.serviceStub;
+        if (!channel.isShutdown()) {
+            channel.shutdownNow();
+            channel.awaitTermination(deadline, TimeUnit.MILLISECONDS);
+        }
 
-            if (this.streamDeadlineMs > 0) {
-                localServiceStub = localServiceStub.withDeadlineAfter(this.streamDeadlineMs, TimeUnit.MILLISECONDS);
-            }
+        this.onConnectionEvent.accept(new ConnectionEvent(false));
+    }
 
-            localServiceStub.eventStream(EventStreamRequest.getDefaultInstance(), responseObserver);
-
-            try {
-                synchronized (sync) {
-                    sync.wait();
-                }
-            } catch (InterruptedException e) {
-                // Interruptions are considered end calls for this observer, hence log and
-                // return
-                // Note - this is the most common interruption when shutdown, hence the log
-                // level debug
-                log.debug("interruption while waiting for condition", e);
-                Thread.currentThread().interrupt();
-            }
+    /**
+     * Handles the event when the GRPC channel becomes ready, marking the connection as established.
+     * Cancels any pending reconnection task and restarts the event stream.
+     */
+    private synchronized void onReady() {
+        connected = true;
 
-            try {
-                backoff.waitUntilNextAttempt();
-            } catch (InterruptedException e) {
-                // Interruptions are considered end calls for this observer, hence log and
-                // return
-                log.warn("interrupted while restarting gRPC Event Stream");
-                Thread.currentThread().interrupt();
-            }
+        if (reconnectTask != null && !reconnectTask.isCancelled()) {
+            reconnectTask.cancel(false);
+            log.info("Reconnection task cancelled as connection became READY.");
         }
+        restartStream();
+    }
 
-        log.error("failed to connect to event stream, exhausted retries");
-        this.onConnectionEvent(false, Collections.emptyList());
+    /**
+     * Handles the event when the GRPC channel loses its connection, marking the connection as lost.
+     * Schedules a reconnection task after a grace period and emits a stale connection event.
+     */
+    private synchronized void onConnectionLost() {
+        log.warn("Connection lost. Emit STALE event...");
```

Comment:
[suggestion] Stale is not something actionable by the sdk user, i do think a info log, or even debug is enough. We should only expect an action by our users, if there is an error

Reply:
I'd argue that loosing the connection (even temporarily) has at least the user facing side effect that evaluations fall back to their default value. Maybe the user want to have a hint why this happened? I interpreted it as a precurser to the error case which is permanently loosing connection. wdyt?

Reply:
I agree that the information is interesting, but more often, this is just information (info) that it happened, and there is no action needed. eg. I know certain companies where error logs will trigger internal processes, and for an expected reconnect within the grace period, a warning would be a high priority.

Reply:
convinced. thx.

Reply:
Another thing to note is that it's possible to attach SDK event handlers to any provider state transition, such as STALE. I'm leaning towards making these all debug for that reason. WDYT?

Reply:
convinced again :)